File: project_euler_unified/problem_278/solution.py
#!/usr/bin/env python3
"""
Problem 278: Linear Combinations of Semiprimes

For distinct primes p, q, r, compute f(p,q,r) = 2pqr - pq - qr - rp.
Sum f over all unordered triples {p,q,r} of distinct primes up to 5000.

Uses prefix sums for O(pi(N)^2) efficiency.
"""

def sieve(limit):
    """Return list of primes up to limit."""
    is_prime = [False, False] + [True] * (limit - 1)
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, limit + 1, i):
                is_prime[j] = False
    return [i for i in range(2, limit + 1) if is_prime[i]], is_prime

def solve():
    LIMIT = 5000
    primes, is_prime = sieve(LIMIT)

    # Prefix sums and counts
    ps = [0] * (LIMIT + 1)
    pc = [0] * (LIMIT + 1)
    for i in range(1, LIMIT + 1):
        ps[i] = ps[i-1] + (i if is_prime[i] else 0)
        pc[i] = pc[i-1] + (1 if is_prime[i] else 0)

    answer = 0
    for i, p in enumerate(primes):
        for j in range(i + 1, len(primes)):
            q = primes[j]

            # r ranges over all primes > q, <= LIMIT
            sum_r = ps[LIMIT] - ps[q]
            count_r = pc[LIMIT] - pc[q]
            if count_r == 0:
                continue

            coeff = 2 * p * q - p - q
            answer += coeff * sum_r - p * q * count_r

    return answer

if __name__ == "__main__":
    answer = solve()
    print(answer)

File: project_euler_unified/problem_278/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 278: Linear Combinations of Semiprimes
 *
 * For distinct primes p, q, r, f(p,q,r) = 2pqr - pq - qr - rp.
 * Sum f over all unordered triples {p,q,r} of distinct primes up to 5000.
 *
 * Optimization: for fixed p < q, sum over all primes r > q (up to 5000):
 *   sum_r f(p,q,r) = (2pq - p - q) * (sum of r) - pq * (count of r)
 *
 * Use prefix sums of primes for O(1) per (p,q) pair.
 * Uses __int128 to handle the large answer (~1.2 * 10^18).
 */

int main(){
    const int LIMIT = 5000;

    vector<bool> is_prime(LIMIT + 1, true);
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i * i <= LIMIT; i++)
        if (is_prime[i])
            for (int j = i*i; j <= LIMIT; j += i)
                is_prime[j] = false;

    vector<int> primes;
    for (int i = 2; i <= LIMIT; i++)
        if (is_prime[i])
            primes.push_back(i);

    // Prefix sums and counts of primes
    vector<long long> ps(LIMIT + 1, 0);
    vector<int> pc(LIMIT + 1, 0);
    for (int i = 1; i <= LIMIT; i++) {
        ps[i] = ps[i-1] + (is_prime[i] ? i : 0);
        pc[i] = pc[i-1] + (is_prime[i] ? 1 : 0);
    }

    __int128 answer = 0;
    int n = primes.size();

    for (int i = 0; i < n; i++) {
        long long p = primes[i];
        for (int j = i + 1; j < n; j++) {
            long long q = primes[j];

            // r ranges over all primes > q and <= LIMIT
            long long sumR = ps[LIMIT] - ps[q];
            long long countR = pc[LIMIT] - pc[q];
            if (countR == 0) continue;

            __int128 coeff = (__int128)2 * p * q - p - q;
            __int128 contribution = coeff * sumR - (__int128)(p * q) * countR;
            answer += contribution;
        }
    }

    // Print __int128
    string s;
    __int128 x = answer;
    if (x == 0) { cout << 0 << endl; return 0; }
    while (x > 0) { s += '0' + (int)(x % 10); x /= 10; }
    reverse(s.begin(), s.end());
    cout << s << endl;

    return 0;
}
